zebra-consensus/src/checkpoint.rs

```diff
@@ -35,7 +45,41 @@ struct CheckpointVerifier<S> {
     /// which only happen in the last few hundred blocks in the chain.
     /// (zcashd allows chain reorganizations up to 99 blocks, and prunes
     /// orphaned side-chains after 288 blocks.)
+    ///
+    /// Checkpoints must start at zero, and have equal spacing. There must
+    /// be at least one checkpoint, the genesis block.
     checkpoint_list: Arc<HashMap<BlockHeight, BlockHeaderHash>>,
+
+    // Derived Variables
+    //
+    /// The height of the maximum checkpoint.
+    ///
+    /// Can be zero, if there is a single genesis block checkpoint.
+    max_checkpoint_height: BlockHeight,
+
+    /// The fixed spacing between checkpoints.
+    ///
+    /// Can be zero, if there is a single genesis block checkpoint.
+    checkpoint_frequency: BlockHeight,
+
+    // Cached Blocks
+    //
+    /// A cache of unverified blocks.
+    ///
+    /// Contains a list of blocks for each checkpoint. The key is the height of
+    /// the ancestor checkpoint. Blocks are verified when there is a chain from
+    /// an ancestor checkpoint to a descendant checkpoint.
+    ///
+    /// Each pair of checkpoints verifies `checkpoint_frequency` blocks, from
+    /// the ancestor checkpoint, to the parent of the descendant checkpoint.
+    /// The final checkpoint does not have any descendant checkpoints, so it
+    /// only verifies a single block.
+    queued: HashMap<BlockHeight, CheckpointInterval>,
+
+    /// The height of the next unverified checkpoint.
```

Comment:
Per the design here: https://github.com/ZcashFoundation/zebra/issues/429#issuecomment-646338626, blocks can come into the checkpointer in any order, but they should resolve in chain order, so that there are no gaps in the blockchain.